File: iosApp/iosApp/Presentation/Main/MyInfo/MyInfoViewModel.swift
//
//  MyInfoViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/06.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class MyInfoViewModel: ObservableObject {
    private let getMyInfoUseCase: GetMyInfoUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let cafeDetailEventPublisher: CafeDetailEventPublisher

    private let castEventPublisher: CastEventPublisher

    private let visitEventPublisher: VisitEventPublisher

    private let userEventPublisher: UserEventPublisher

    @Published private(set) var uiState = MyInfoUiState.empty

    let event = PassthroughSubject<MyInfoEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func observeSession() {
        tasks[.session]?.cancel()
        tasks[.session] = Task {
            do {
                for try await _ in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    self.loadMyInfo()
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }
    
    private func loadMyInfo() {
        tasks[.loadMyInfo]?.cancel()

        uiState.isLoading = true
        uiState.errorMessage = nil

        tasks[.loadMyInfo] = Task {
            do {
                let result = try await getMyInfoUseCase.invoke()

                if let success = result as? AppResultSuccess<AnyObject>,
                let feed = success.data as? Shared.MyInfoFeed {
                    let normalizedRecentVisits = normalizeCafes(
                        feed.recentVisits,
                        maxCount: feed.recentVisits.count
                    )
                    let normalizedFavorites = normalizeCafes(
                        feed.favorites,
                        maxCount: feed.favorites.count
                    )
                    uiState = MyInfoUiState(
                        isLoading: false,
                        errorMessage: nil,
                        isLoggedIn: feed.isLoggedIn,
                        user: feed.user,
                        summary: feed.summary,
                        castDetail: feed.castDetail,
                        ownedCafes: feed.ownedCafes,
                        badges: feed.badges,
                        popularCafes: feed.popularCafes,
                        recentVisits: normalizedRecentVisits,
                        favorites: normalizedFavorites,
                        followedMaids: feed.followedMaids,
                        isLoginPromptVisible: false
                    )
                } else if let failure = result as? AppResultFailure {
                    uiState = .empty
                    uiState.errorMessage = "\(failure.error)"
                } else {
                    uiState = .empty
                }
            } catch {
                if Task.isCancelled { return }

                uiState = .empty
                uiState.errorMessage = error.localizedDescription
            }
        }
    }

    private func observeCafeDetailEvent() {
        tasks[.cafeDetailEvent]?.cancel()
        tasks[.cafeDetailEvent] = Task {
            do {
                for try await event in asyncSequence(for: cafeDetailEventPublisher.events) {
                    if let updated = event as? CafeDetailEvent.CafeInfoUpdated {
                        self.patchCafe(updated.cafe)
                    } else if event is CafeDetailEvent.FavoriteToggled {
                        self.loadMyInfo()
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeVisitEvent() {
        tasks[.visitEvent]?.cancel()
        tasks[.visitEvent] = Task {
            do {
                for try await event in asyncSequence(for: visitEventPublisher.events) {
                    switch event {
                    case is VisitEvent.Created:
                        self.applyVisitCountDelta(1)
                    case is VisitEvent.Deleted:
                        self.applyVisitCountDelta(-1)
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeCastEvent() {
        tasks[.castEvent]?.cancel()
        tasks[.castEvent] = Task {
            do {
                for try await event in asyncSequence(for: castEventPublisher.events) {
                    switch event {
                    case let updated as Shared.CastEvent.Updated:
                        if let isFollowing = updated.isFollowing?.boolValue {
                            self.updateFollowedCast(updated.cast, isFollowing: isFollowing)
                        } else {
                            self.patchCast(updated.cast)
                        }
                    case let deleted as Shared.CastEvent.Deleted:
                        self.removeCast(deleted.castId)
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeUserEvent() {
        tasks[.userEvent]?.cancel()
        tasks[.userEvent] = Task {
            do {
                for try await event in asyncSequence(for: userEventPublisher.events) {
                    switch event {
                    case let updated as Shared.UserEvent.ProfileUpdated:
                        self.patchUser(updated.user)
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func patchCafe(_ cafe: Cafe) {
        uiState.ownedCafes = uiState.ownedCafes.map { item in
            guard item.id == cafe.id else { return item }
            return CafeManagementData.OwnedCafeSummary(
                id: item.id,
                name: cafe.name,
                city: cafe.region.city,
                isApproved: item.isApproved,
                todayVisitors: item.todayVisitors,
                todayCheckIns: item.todayCheckIns,
                todayReviews: item.todayReviews,
                rating: cafe.ratingAvg,
                castCount: item.castCount,
                noticeCount: item.noticeCount,
                externalLinkCount: item.externalLinkCount,
                thumbnailImage: cafe.thumbnailImage
            )
        }
        uiState.popularCafes = uiState.popularCafes.map { $0.id == cafe.id ? cafe : $0 }
        uiState.recentVisits = uiState.recentVisits.map { $0.id == cafe.id ? cafe : $0 }
        uiState.favorites = uiState.favorites.map { $0.id == cafe.id ? cafe : $0 }
        if let detail = uiState.castDetail, detail.cafe.id == cafe.id {
            uiState.castDetail = CastDetail(
                cast: detail.cast,
                cafe: cafe,
                images: detail.images,
                schedule: detail.schedule,
                visitCertificationCount: detail.visitCertificationCount
            )
        }
    }

    private func patchCast(_ cast: Cast) {
        uiState.followedMaids = uiState.followedMaids.map { $0.id == cast.id ? cast : $0 }
        if let detail = uiState.castDetail, detail.cast.id == cast.id {
            uiState.castDetail = CastDetail(
                cast: cast,
                cafe: detail.cafe,
                images: detail.images,
                schedule: detail.schedule,
                visitCertificationCount: detail.visitCertificationCount
            )
        }
    }

    private func removeCast(_ castId: String) {
        uiState.followedMaids.removeAll { $0.id == castId }
        if uiState.castDetail?.cast.id == castId {
            uiState.castDetail = nil
        }
    }

    private func patchUser(_ user: User) {
        if uiState.user?.id == user.id {
            uiState.user = user
        }
    }

    private func updateFollowedCast(_ cast: Cast, isFollowing: Bool) {
        if isFollowing {
            uiState.followedMaids = upsertFollowedCast(uiState.followedMaids, cast: cast)
        } else {
            uiState.followedMaids.removeAll { $0.id == cast.id }
        }
        if let summary = uiState.summary {
            uiState.summary = MyPageSummary(
                userId: summary.userId,
                totalVisits: summary.totalVisits,
                favoritesCount: summary.favoritesCount,
                followedCastsCount: Int32(uiState.followedMaids.count),
                badgesCount: summary.badgesCount,
                level: summary.level
            )
        }
        if let detail = uiState.castDetail, detail.cast.id == cast.id {
            uiState.castDetail = CastDetail(
                cast: cast,
                cafe: detail.cafe,
                images: detail.images,
                schedule: detail.schedule,
                visitCertificationCount: detail.visitCertificationCount
            )
        }
    }

    private func upsertFollowedCast(_ items: [Cast], cast: Cast) -> [Cast] {
        var result = items
        if let index = result.firstIndex(where: { $0.id == cast.id }) {
            result[index] = cast
        } else {
            result.insert(cast, at: 0)
        }
        return result
    }

    private func applyVisitCountDelta(_ delta: Int) {
        guard delta != 0, let summary = uiState.summary else { return }

        let nextVisitCount = max(Int(summary.totalVisits) + delta, 0)
        let nextStampCount = max(Int(summary.badgesCount) + delta, 0)
        let nextLevel = max(1, 1 + (nextVisitCount / 5))
        uiState.summary = MyPageSummary(
            userId: summary.userId,
            totalVisits: Int32(nextVisitCount),
            favoritesCount: summary.favoritesCount,
            followedCastsCount: summary.followedCastsCount,
            badgesCount: Int32(nextStampCount),
            level: Int32(nextLevel)
        )
        let favoritesCount = Int(summary.favoritesCount)
        let followedCount = Int(summary.followedCastsCount)
        uiState.badges = uiState.badges.map { badge in
            let unlocked: Bool

            switch badge.id {
            case "badge-checkin-starter":
                unlocked = nextStampCount >= 1
            case "badge-stamp-collector":
                unlocked = nextStampCount >= 3
            case "badge-regular-visitor":
                unlocked = nextVisitCount >= 5
            case "badge-checkin-veteran":
                unlocked = nextVisitCount >= 10
            case "badge-favorite-curator":
                unlocked = favoritesCount >= 3
            case "badge-favorite-master":
                unlocked = favoritesCount >= 10
            case "badge-cast-supporter":
                unlocked = followedCount >= 3
            case "badge-cast-ambassador":
                unlocked = followedCount >= 10
            case "badge-level-up":
                unlocked = nextLevel >= 3
            case "badge-concafe-master":
                unlocked = nextStampCount >= 10
            default:
                unlocked = badge.unlocked
            }
            return ProfileBadge(
                id: badge.id,
                name: badge.name,
                icon: badge.icon,
                unlocked: unlocked
            )
        }
    }

    func onAction(_ action: MyInfoAction) {
        switch action {
        case .cafeTapped(let id):
            if uiState.isLoggedIn {
                event.send(.navigateToCafe(id: id))
            } else {
                uiState.isLoginPromptVisible = true
            }
        case .maidTapped(let id):
            if uiState.isLoggedIn {
                event.send(.navigateToCast(id: id))
            } else {
                uiState.isLoginPromptVisible = true
            }
        case .loginPromptSignInTapped:
            uiState.isLoginPromptVisible = false
            event.send(.navigateToSignIn)
        case .dismissLoginPrompt:
            uiState.isLoginPromptVisible = false
        case .signInTapped:
            event.send(.navigateToSignIn)
        case .refresh:
            loadMyInfo()
        }
    }

    init(
        getMyInfoUseCase: GetMyInfoUseCase = KoinInitializerKt.resolveGetMyInfoUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        cafeDetailEventPublisher: CafeDetailEventPublisher = KoinInitializerKt.resolveCafeDetailEventPublisher(),
        castEventPublisher: CastEventPublisher = KoinInitializerKt.resolveCastEventPublisher(),
        visitEventPublisher: VisitEventPublisher = KoinInitializerKt.resolveVisitEventPublisher(),
        userEventPublisher: UserEventPublisher = KoinInitializerKt.resolveUserEventPublisher()
    ) {
        self.getMyInfoUseCase = getMyInfoUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.cafeDetailEventPublisher = cafeDetailEventPublisher
        self.castEventPublisher = castEventPublisher
        self.visitEventPublisher = visitEventPublisher
        self.userEventPublisher = userEventPublisher

        observeSession()
        observeCafeDetailEvent()
        observeCastEvent()
        observeVisitEvent()
        observeUserEvent()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case loadMyInfo
        case session
        case cafeDetailEvent
        case castEvent
        case visitEvent
        case userEvent
    }

    private func normalizeCafes(_ cafes: [Cafe], maxCount: Int) -> [Cafe] {
        var seen = Set<String>()
        var normalized: [Cafe] = []

        for cafe in cafes {
            if seen.contains(cafe.id) { continue }
            seen.insert(cafe.id)
            normalized.append(cafe)
            if maxCount > 0 && normalized.count >= maxCount {
                break
            }
        }
        return normalized
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/myinfo/MyInfoViewModel.kt
package com.hhp227.concafe.presentation.main.myinfo

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.Cafe
import com.hhp227.concafe.domain.model.CastSchedule
import com.hhp227.concafe.domain.model.CastScheduleStatus
import com.hhp227.concafe.domain.model.MyPageSummary
import com.hhp227.concafe.domain.model.User
import com.hhp227.concafe.domain.event.CafeDetailEvent
import com.hhp227.concafe.domain.model.Cast
import com.hhp227.concafe.domain.event.CastEvent
import com.hhp227.concafe.domain.event.ScheduleManagementEvent
import com.hhp227.concafe.domain.event.UserEvent
import com.hhp227.concafe.domain.event.VisitEvent
import com.hhp227.concafe.domain.event.publisher.CafeDetailEventPublisher
import com.hhp227.concafe.domain.event.publisher.CastEventPublisher
import com.hhp227.concafe.domain.event.publisher.ScheduleManagementEventPublisher
import com.hhp227.concafe.domain.event.publisher.UserEventPublisher
import com.hhp227.concafe.domain.event.publisher.VisitEventPublisher
import com.hhp227.concafe.domain.usecase.GetMyInfoUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase
import com.hhp227.concafe.presentation.main.myinfo.MyInfoEvent.*
import com.hhp227.concafe.presentation.main.myinfo.MyInfoUiState.Companion.empty
import kotlin.math.max

class MyInfoViewModel(
    private val getMyInfoUseCase: GetMyInfoUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase,
    private val cafeDetailEventPublisher: CafeDetailEventPublisher,
    private val castEventPublisher: CastEventPublisher,
    private val visitEventPublisher: VisitEventPublisher,
    private val userEventPublisher: UserEventPublisher,
    private val scheduleManagementEventPublisher: ScheduleManagementEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(empty())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<MyInfoEvent>()
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<TaskKey, Job>()

    private fun observeSession() {
        viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest { user ->
                _uiState.update {
                    it.copy(
                        isLoggedIn = user != null,
                        user = user
                    )
                }
                loadMyInfo()
            }
        }
    }

    private fun loadMyInfo() {
        jobs[TaskKey.LOAD_MY_INFO]?.cancel()
        jobs[TaskKey.LOAD_MY_INFO] = viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            when (val result = getMyInfoUseCase.invoke()) {
                is AppResult.Success -> {
                    val normalizedRecentVisits = normalizeCafes(
                        items = result.data.recentVisits,
                        maxCount = result.data.recentVisits.size
                    )
                    val normalizedFavorites = normalizeCafes(
                        items = result.data.favorites,
                        maxCount = result.data.favorites.size
                    )
                    _uiState.value = MyInfoUiState(
                        isLoading = false,
                        errorMessage = null,
                        isLoggedIn = result.data.isLoggedIn,
                        user = result.data.user,
                        summary = result.data.summary,
                        castDetail = result.data.castDetail,
                        ownedCafes = result.data.ownedCafes,
                        badges = result.data.badges,
                        popularCafes = result.data.popularCafes,
                        recentVisits = normalizedRecentVisits,
                        favorites = normalizedFavorites,
                        followedMaids = result.data.followedMaids,
                        isLoginPromptVisible = false
                    )
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    private fun observeCafeDetailEvent() {
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT] = viewModelScope.launch {
            cafeDetailEventPublisher.events.collectLatest { event ->
                when (event) {
                    is CafeDetailEvent.CafeInfoUpdated -> patchCafe(event.cafe)
                    is CafeDetailEvent.FavoriteToggled -> {
                        if (event.isFavorite) {
                            refreshFavoritesSection()
                        } else {
                            removeFavoriteCafe(event.cafeId)
                        }
                    }
                    is CafeDetailEvent.MenuCreated,
                    is CafeDetailEvent.MenuUpdated,
                    is CafeDetailEvent.MenuDeleted,
                    is CafeDetailEvent.GoodsCreated,
                    is CafeDetailEvent.GoodsUpdated,
                    is CafeDetailEvent.GoodsDeleted -> Unit
                }
            }
        }
    }

    private fun observeVisitEvent() {
        jobs[TaskKey.OBSERVE_VISIT_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_VISIT_EVENT] = viewModelScope.launch {
            visitEventPublisher.events.collectLatest { event ->
                when (event) {
                    is VisitEvent.Created -> {
                        applyVisitCountDelta(1)
                        refreshRecentVisitsSection()
                    }
                    is VisitEvent.Deleted -> {
                        applyVisitCountDelta(-1)
                        refreshRecentVisitsSection()
                    }
                }
            }
        }
    }

    private fun refreshRecentVisitsSection() {
        jobs[TaskKey.REFRESH_RECENT_VISITS]?.cancel()
        jobs[TaskKey.REFRESH_RECENT_VISITS] = viewModelScope.launch {
            when (val result = getMyInfoUseCase.invoke()) {
                is AppResult.Success -> {
                    val normalizedRecentVisits = normalizeCafes(
                        items = result.data.recentVisits,
                        maxCount = result.data.recentVisits.size
                    )

                    _uiState.update { state ->
                        state.copy(recentVisits = normalizedRecentVisits)
                    }
                }
                is AppResult.Failure -> Unit
            }
        }
    }

    private fun observeCastEvent() {
        jobs[TaskKey.OBSERVE_CAST_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAST_EVENT] = viewModelScope.launch {
            castEventPublisher.events.collectLatest { event ->
                when (event) {
                    is CastEvent.Created -> Unit
                    is CastEvent.Updated -> {
                        val isFollowing = event.isFollowing

                        if (isFollowing != null) {
                            updateFollowedCast(event.cast, isFollowing)
                        } else {
                            patchCast(event.cast)
                        }
                    }
                    is CastEvent.Deleted -> removeCast(event.castId)
                }
            }
        }
    }

    private fun observeUserEvent() {
        jobs[TaskKey.OBSERVE_USER_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_USER_EVENT] = viewModelScope.launch {
            userEventPublisher.events.collectLatest { event ->
                when (event) {
                    is UserEvent.ProfileUpdated -> patchUser(event.user)
                }
            }
        }
    }

    private fun observeScheduleManagementEvent() {
        jobs[TaskKey.OBSERVE_SCHEDULE_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_SCHEDULE_EVENT] = viewModelScope.launch {
            scheduleManagementEventPublisher.events.collectLatest { event ->
                when (event) {
                    is ScheduleManagementEvent.Updated -> patchCastSchedule(event)
                }
            }
        }
    }

    private fun patchCastSchedule(event: ScheduleManagementEvent.Updated) {
        val detail = _uiState.value.castDetail ?: return
        if (detail.cast.id != event.castId) return
        val updatedSchedule = when (event.status) {
            CastScheduleStatus.WORK -> {
                val startTime = event.startTime ?: return
                val endTime = event.endTime ?: return
                val existing = detail.schedule.firstOrNull { it.date == event.date }
                val newEntry = CastSchedule(
                    id = existing?.id ?: "schedule-${event.castId}-${event.date.replace("-", "")}",
                    castId = event.castId,
                    cafeId = detail.cafe.id,
                    date = event.date,
                    startTime = startTime,
                    endTime = endTime
                )

                detail.schedule.filterNot { it.date == event.date } + newEntry
            }
            CastScheduleStatus.OFF,
            CastScheduleStatus.VACATION -> detail.schedule.filterNot { it.date == event.date }
        }
        _uiState.update { it.copy(castDetail = detail.copy(schedule = updatedSchedule)) }
    }

    private fun patchCafe(cafe: Cafe) {
        _uiState.update { state ->
            state.copy(
                castDetail = state.castDetail?.takeIf { it.cafe.id == cafe.id }?.copy(cafe = cafe) ?: state.castDetail,
                ownedCafes = state.ownedCafes.map { item ->
                    if (item.id == cafe.id) {
                        item.copy(name = cafe.name, city = cafe.region.city, rating = cafe.ratingAvg)
                    } else {
                        item
                    }
                },
                popularCafes = state.popularCafes.map { item -> if (item.id == cafe.id) cafe else item },
                recentVisits = state.recentVisits.map { item -> if (item.id == cafe.id) cafe else item },
                favorites = state.favorites.map { item -> if (item.id == cafe.id) cafe else item }
            )
        }
    }

    private fun removeFavoriteCafe(cafeId: String) {
        _uiState.update { state ->
            val hadFavorite = state.favorites.any { item -> item.id == cafeId }
            if (!hadFavorite) {
                state
            } else {
                val nextFavorites = state.favorites.filterNot { item -> item.id == cafeId }
                val nextSummary = state.summary?.let { summary ->
                    summary.copy(favoritesCount = max(summary.favoritesCount - 1, 0))
                }
                val totalVisits = nextSummary?.totalVisits ?: state.summary?.totalVisits ?: 0
                val favoritesCount = nextSummary?.favoritesCount ?: state.summary?.favoritesCount ?: 0
                val followedCount = nextSummary?.followedCastsCount ?: state.summary?.followedCastsCount ?: 0
                val badgesCount = nextSummary?.badgesCount ?: state.summary?.badgesCount ?: 0
                val level = nextSummary?.level ?: state.summary?.level ?: 1

                state.copy(
                    summary = nextSummary,
                    favorites = nextFavorites,
                    badges = state.badges.map { badge ->
                        val isUnlocked = when (badge.id) {
                            "badge-checkin-starter" -> badgesCount >= 1
                            "badge-stamp-collector" -> badgesCount >= 3
                            "badge-regular-visitor" -> totalVisits >= 5
                            "badge-checkin-veteran" -> totalVisits >= 10
                            "badge-favorite-curator" -> favoritesCount >= 3
                            "badge-favorite-master" -> favoritesCount >= 10
                            "badge-cast-supporter" -> followedCount >= 3
                            "badge-cast-ambassador" -> followedCount >= 10
                            "badge-level-up" -> level >= 3
                            "badge-concafe-master" -> badgesCount >= 10
                            else -> badge.unlocked
                        }
                        badge.copy(unlocked = isUnlocked)
                    }
                )
            }
        }
    }

    private fun refreshFavoritesSection() {
        jobs[TaskKey.REFRESH_FAVORITES]?.cancel()
        jobs[TaskKey.REFRESH_FAVORITES] = viewModelScope.launch {
            when (val result = getMyInfoUseCase.invoke()) {
                is AppResult.Success -> {
                    val normalizedFavorites = normalizeCafes(
                        items = result.data.favorites,
                        maxCount = result.data.favorites.size
                    )
                    _uiState.update { state ->
                        state.copy(
                            summary = result.data.summary,
                            badges = result.data.badges,
                            favorites = normalizedFavorites
                        )
                    }
                }
                is AppResult.Failure -> Unit
            }
        }
    }

    private fun patchCast(cast: Cast) {
        _uiState.update { state ->
            state.copy(
                castDetail = state.castDetail?.takeIf { it.cast.id == cast.id }?.copy(cast = cast) ?: state.castDetail,
                followedMaids = state.followedMaids.map { item -> if (item.id == cast.id) cast else item }
            )
        }
    }

    private fun removeCast(castId: String) {
        _uiState.update { state ->
            state.copy(
                castDetail = state.castDetail?.takeUnless { it.cast.id == castId },
                followedMaids = state.followedMaids.filterNot { it.id == castId }
            )
        }
    }

    private fun patchUser(user: User) {
        _uiState.update { state ->
            if (state.user?.id == user.id) {
                state.copy(user = user)
            } else {
                state
            }
        }
    }

    private fun updateFollowedCast(cast: Cast, isFollowing: Boolean) {
        _uiState.update { state ->
            val followedMaids = if (isFollowing) {
                upsertFollowedCast(state.followedMaids, cast)
            } else {
                state.followedMaids.filterNot { item -> item.id == cast.id }
            }
            val summary = updateFollowedCount(
                summary = state.summary,
                followedMaids = followedMaids
            )
            state.copy(
                castDetail = state.castDetail?.takeIf { detail -> detail.cast.id == cast.id }?.copy(cast = cast)
                    ?: state.castDetail,
                summary = summary,
                followedMaids = followedMaids
            )
        }
    }

    private fun upsertFollowedCast(items: List<Cast>, cast: Cast): List<Cast> {
        val hasCast = items.any { item -> item.id == cast.id }
        return if (hasCast) {
            items.map { item -> if (item.id == cast.id) cast else item }
        } else {
            listOf(cast) + items
        }
    }

    private fun updateFollowedCount(
        summary: MyPageSummary?,
        followedMaids: List<Cast>
    ): MyPageSummary? {
        return summary?.copy(followedCastsCount = followedMaids.size)
    }

    private fun applyVisitCountDelta(delta: Int) {
        if (delta == 0) return
        _uiState.update { state ->
            val summary = state.summary

            if (summary == null) {
                state
            } else {
                val nextVisitCount = max(summary.totalVisits + delta, 0)
                val nextStampCount = max(summary.badgesCount + delta, 0)
                val nextLevel = max(1, 1 + (nextVisitCount / 5))
                val nextSummary = summary.copy(
                    totalVisits = nextVisitCount,
                    badgesCount = nextStampCount,
                    level = nextLevel
                )
                val favoritesCount = nextSummary.favoritesCount
                val followedCount = nextSummary.followedCastsCount
                state.copy(
                    summary = nextSummary,
                    badges = state.badges.map { badge ->
                        val isUnlocked = when (badge.id) {
                            "badge-checkin-starter" -> nextStampCount >= 1
                            "badge-stamp-collector" -> nextStampCount >= 3
                            "badge-regular-visitor" -> nextVisitCount >= 5
                            "badge-checkin-veteran" -> nextVisitCount >= 10
                            "badge-favorite-curator" -> favoritesCount >= 3
                            "badge-favorite-master" -> favoritesCount >= 10
                            "badge-cast-supporter" -> followedCount >= 3
                            "badge-cast-ambassador" -> followedCount >= 10
                            "badge-level-up" -> nextLevel >= 3
                            "badge-concafe-master" -> nextStampCount >= 10
                            else -> badge.unlocked
                        }
                        badge.copy(unlocked = isUnlocked)
                    }
                )
            }
        }
    }

    fun onAction(action: MyInfoAction) {
        when (action) {
            is MyInfoAction.ClickCafe -> viewModelScope.launch {
                if (_uiState.value.isLoggedIn) {
                    _event.emit(NavigateToCafe(action.id))
                } else {
                    _uiState.update { it.copy(isLoginPromptVisible = true) }
                }
            }
            is MyInfoAction.ClickMaid -> viewModelScope.launch {
                if (_uiState.value.isLoggedIn) {
                    _event.emit(NavigateToCast(action.id))
                } else {
                    _uiState.update { it.copy(isLoginPromptVisible = true) }
                }
            }
            MyInfoAction.ClickLoginPromptSignIn -> viewModelScope.launch {
                _uiState.update { it.copy(isLoginPromptVisible = false) }
                _event.emit(NavigateToSignIn)
            }
            MyInfoAction.DismissLoginPrompt -> {
                _uiState.update { it.copy(isLoginPromptVisible = false) }
            }
            MyInfoAction.ClickSignIn -> viewModelScope.launch {
                _event.emit(NavigateToSignIn)
            }
            MyInfoAction.Refresh -> loadMyInfo()
        }
    }

    init {
        observeSession()
        observeCafeDetailEvent()
        observeCastEvent()
        observeVisitEvent()
        observeUserEvent()
        observeScheduleManagementEvent()
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        super.onCleared()
    }

    private enum class TaskKey {
        LOAD_MY_INFO,
        REFRESH_RECENT_VISITS,
        REFRESH_FAVORITES,
        OBSERVE_CAFE_DETAIL_EVENT,
        OBSERVE_CAST_EVENT,
        OBSERVE_VISIT_EVENT,
        OBSERVE_USER_EVENT,
        OBSERVE_SCHEDULE_EVENT
    }

    private fun normalizeCafes(items: List<Cafe>, maxCount: Int): List<Cafe> {
        val normalizedItems = items
            .distinctBy { it.id }
        return if (maxCount <= 0) {
            normalizedItems
        } else {
            normalizedItems.take(maxCount)
        }
    }
}
